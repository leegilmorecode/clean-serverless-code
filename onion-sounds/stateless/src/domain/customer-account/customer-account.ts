import {
  CreateCustomerAccountProps,
  NewCustomerAccountProps,
  PaymentStatus,
  SubscriptionType,
  UnmarshalledCustomerAccount,
} from '@models/types';

import { Entity } from '@entity/entity';
import { PaymentInvalidError } from '@errors/payment-invalid-error';
import { schema } from '@schemas/customer-account.schema';

// only works with entities and value objects - calling the repositories
// adapter --> use case --> (domain) <-- repository <-- adapter

/**
 * Customer Account Domain Entity
 *
 * A customer account is an entity which allows customers to create a new account, and upgrade their
 * account subscription.
 */
export class CustomerAccount extends Entity<CreateCustomerAccountProps> {
  private constructor({
    id,
    created,
    updated,
    ...props
  }: CreateCustomerAccountProps) {
    super(props, id, created, updated);
  }

  // create an instance of a new customer account through a factory method
  // adapter --> use case --> (domain)
  public static createAccount(props: NewCustomerAccountProps): CustomerAccount {
    const customerAccountProps: CreateCustomerAccountProps = {
      firstName: props.firstName,
      surname: props.surname,
      subscriptionType: SubscriptionType.Basic,
      paymentStatus: PaymentStatus.Valid,
    };

    const instance: CustomerAccount = new CustomerAccount(customerAccountProps);
    instance.validate(schema);

    return instance;
  }

  // upgrade a customer account from basic to advanced through a service (use case)
  // adapter --> use case --> (domain)
  public upgradeAccount(): void {
    // only allow an upgrade if the payment status is valid
    if (this.props.paymentStatus === PaymentStatus.Invalid) {
      throw new PaymentInvalidError('Payment is invalid - unable to upgrade');
    }

    // update the account to upgraded
    this.props.subscriptionType = SubscriptionType.Upgraded;

    this.setUpdatedDate();
    this.validate(schema);
  }

  // create a dto based on the domain instance
  public toDto(): UnmarshalledCustomerAccount {
    return {
      id: this.id,
      created: this.created,
      updated: this.updated,
      firstName: this.props.firstName,
      surname: this.props.surname,
      subscriptionType: this.props.subscriptionType,
      paymentStatus: this.props.paymentStatus,
    };
  }

  // create a domain object based on the dto
  public static toDomain(raw: UnmarshalledCustomerAccount): CustomerAccount {
    const instance = new CustomerAccount(raw);
    return instance;
  }
}
