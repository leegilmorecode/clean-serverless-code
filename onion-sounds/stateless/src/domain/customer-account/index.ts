export { CustomerAccount } from './customer-account';
