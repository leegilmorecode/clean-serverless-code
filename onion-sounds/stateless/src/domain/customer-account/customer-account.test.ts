import {
  CustomerAccountProps,
  NewCustomerAccountProps,
  PaymentStatus,
  SubscriptionType,
  UnmarshalledCustomerAccount,
} from '@models/types';

import { CustomerAccount } from '@domain/customer-account';

let customer: UnmarshalledCustomerAccount = {
  created: '2022-01-01T00:00:00.000Z',
  firstName: 'Lee',
  id: 'f39e49ad-8f88-448f-8a15-41d560ad6d70',
  paymentStatus: PaymentStatus.Valid,
  subscriptionType: SubscriptionType.Basic,
  surname: 'Gilmore',
  updated: '2022-01-01T00:00:00.000Z',
};

describe('customer-account', () => {
  beforeAll(() => {
    jest.useFakeTimers().setSystemTime(new Date('2022-01-01'));
  });

  afterAll(() => {
    jest.clearAllTimers();
    jest.clearAllMocks();
  });

  describe('create account', () => {
    it('should fail creating a new account with an invalid customer', () => {
      // arrange
      const newCustomerAccount: NewCustomerAccountProps = {
        firstName: 'Â±', // invalid
        surname: 'Gilmore',
      };

      // act / assert
      expect(() =>
        CustomerAccount.createAccount(newCustomerAccount)
      ).toThrowErrorMatchingInlineSnapshot(
        `"[{\\"instancePath\\":\\"/firstName\\",\\"schemaPath\\":\\"#/properties/firstName/pattern\\",\\"keyword\\":\\"pattern\\",\\"params\\":{\\"pattern\\":\\"^[a-zA-Z]+$\\"},\\"message\\":\\"must match pattern \\\\\\"^[a-zA-Z]+$\\\\\\"\\"}]"`
      );
    });

    it('should create the new account successfully with a valid customer', () => {
      // arrange
      const newCustomerAccount: NewCustomerAccountProps = {
        firstName: 'Lee',
        surname: 'Gilmore',
      };

      // act
      const newCustomer = CustomerAccount.createAccount(newCustomerAccount);

      // assert
      expect(newCustomer).toMatchSnapshot();
    });
  });

  describe('upgrade account', () => {
    it('should upgrade an existing customer by id if payment status is valid', () => {
      // arrange
      const customerAccount: CustomerAccount = CustomerAccount.toDomain({
        ...customer,
        subscriptionType: SubscriptionType.Upgraded,
      });

      // act
      customerAccount.upgradeAccount();

      // assert
      expect(customerAccount).toMatchSnapshot();
    });

    it('should not upgrade an existing customer by id if payment status is invalid', () => {
      // arrange
      const customerAccount: CustomerAccount = CustomerAccount.toDomain({
        ...customer,
        paymentStatus: PaymentStatus.Invalid,
      });

      // act / assert
      expect(() => customerAccount.upgradeAccount()).toThrow(
        'Payment is invalid - unable to upgrade'
      );
    });
  });

  describe('toDto', () => {
    it('should create the correct dto', () => {
      // arrange
      const newCustomerAccount: NewCustomerAccountProps = {
        firstName: 'Lee',
        surname: 'Gilmore',
      };

      // act
      const newCustomer = CustomerAccount.createAccount(newCustomerAccount);

      // assert
      expect(newCustomer.toDto()).toMatchSnapshot();
    });
  });

  describe('toDomain', () => {
    it('should create a domain object based on a dto', () => {
      // arrange
      const customerAccountProps: CustomerAccountProps = {
        id: 'f39e49ad-8f88-448f-8a15-41d560ad6d70',
        firstName: 'Lee',
        surname: 'Gilmore',
        paymentStatus: PaymentStatus.Valid,
        subscriptionType: SubscriptionType.Upgraded,
        created: '2022-01-01T00:00:00.000Z',
        updated: '2022-01-01T00:00:00.000Z',
      };

      // act
      const customerAccount: CustomerAccount =
        CustomerAccount.toDomain(customerAccountProps);

      // assert
      expect(customerAccount).toMatchSnapshot();
    });
  });
});
