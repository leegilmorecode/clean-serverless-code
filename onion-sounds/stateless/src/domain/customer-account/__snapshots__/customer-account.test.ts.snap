// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customer-account create account should create the new account successfully with a valid customer 1`] = `
CustomerAccount {
  "_created": "2022-01-01T00:00:00.000Z",
  "_id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
  "_updated": "2022-01-01T00:00:00.000Z",
  "props": Object {
    "created": "2022-01-01T00:00:00.000Z",
    "firstName": "Lee",
    "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
    "paymentStatus": "Valid",
    "subscriptionType": "Basic",
    "surname": "Gilmore",
    "updated": "2022-01-01T00:00:00.000Z",
  },
}
`;

exports[`customer-account toDomain should create a domain object based on a dto 1`] = `
CustomerAccount {
  "_created": "2022-01-01T00:00:00.000Z",
  "_id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
  "_updated": "2022-01-01T00:00:00.000Z",
  "props": Object {
    "created": "2022-01-01T00:00:00.000Z",
    "firstName": "Lee",
    "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
    "paymentStatus": "Valid",
    "subscriptionType": "Upgraded",
    "surname": "Gilmore",
    "updated": "2022-01-01T00:00:00.000Z",
  },
}
`;

exports[`customer-account toDto should create the correct dto 1`] = `
Object {
  "created": "2022-01-01T00:00:00.000Z",
  "firstName": "Lee",
  "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
  "paymentStatus": "Valid",
  "subscriptionType": "Basic",
  "surname": "Gilmore",
  "updated": "2022-01-01T00:00:00.000Z",
}
`;

exports[`customer-account upgrade account should upgrade an existing customer by id if payment status is valid 1`] = `
CustomerAccount {
  "_created": "2022-01-01T00:00:00.000Z",
  "_id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
  "_updated": "2022-01-01T00:00:00.000Z",
  "props": Object {
    "created": "2022-01-01T00:00:00.000Z",
    "firstName": "Lee",
    "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
    "paymentStatus": "Valid",
    "subscriptionType": "Upgraded",
    "surname": "Gilmore",
    "updated": "2022-01-01T00:00:00.000Z",
  },
}
`;
